### ISSUE TITLE
FasterHashing crash on FreeBSD

### ISSUE THREAD

**Issue by buengese** [duplicati/duplicati#2892]
- [ x] I have searched open and closed issues for duplicates.

----------------------------------------

## Environment info
<!-- Please include some relevant information about your environment. -->
<!-- For "Backend", please indicate the backup destination (e.g. Amazon S3, OneDrive, FTP, WebDAV, local). -->
- **Duplicati version**: >= v2.0.2.9-2.0.2.9_canary_2017-10-01
- **Operating system**: FreeBSD 11.0 - Mono 5.0.1
- **Backend**: Any

## Description
<!-- Describe the issue that you are experiencing below. -->
I'm a little late on this one. The FasterHashing library causes Mono to segfault on FreeBSD. Easy fix would be to just disable usage of FasterHashing on FreeBSD and fall back to the old implementation.

## log
<!-- Posting a debug log helps the developers find and fix your particular issue more easily. -->
<!-- Please wrap your code in code blocks with triple back-ticks to increase readability. -->
```
Stacktrace:                                                                                                                         
                                                                                                                                    
  at <unknown> <0xffffffff>                                                                                                         
  at (wrapper managed-to-native) FasterHashing.InteropOpenSSL10_libssl.OpenSSL_add_all_digests () [0x00002] in <787629f3dc434608a1b2
bf412946d433>:0                                                                                                                     
  at FasterHashing.OpenSSL10_libssl_HashAlgorithmSHA256..ctor () [0x0000d] in <787629f3dc434608a1b2bf412946d433>:0                  
  at FasterHashing.OpenSSL10_libssl_HashAlgorithm.Create (string) [0x00036] in <787629f3dc434608a1b2bf412946d433>:0                 
  at FasterHashing.OpenSSL10HashAlgorithm.Create (string,FasterHashing.OpenSSL10HashAlgorithm/LibraryName) [0x00035] in <787629f3dc4
34608a1b2bf412946d433>:0                                                                                                            
  at FasterHashing.FasterHash.Create (string,bool,FasterHashing.HashImplementation) [0x0002a] in <787629f3dc434608a1b2bf412946d433>:
0                                                                                                                                   
  at Duplicati.Library.Utility.HashAlgorithmHelper.Create (string) [0x00000] in <9e3565512d4d4bb39442d539c512c851>:0                
  at Duplicati.Library.Main.BackendManager.CalculateFileHash (System.IO.Stream) [0x00000] in <82db7c3af7d944569906ea5ff7f18985>:0   
  at Duplicati.Library.Main.BackendManager/<>c__DisplayClass34_0.<coreDoGetPiping>b__1 () [0x0000c] in <82db7c3af7d944569906ea5ff7f1
8985>:0                                                                                                                             
  at System.Threading.Tasks.Task`1<TResult_REF>.InnerInvoke () [0x0000f] in <a9947af07e614fee8d8cc4b107980e51>:0                    
  at System.Threading.Tasks.Task.Execute () [0x00010] in <a9947af07e614fee8d8cc4b107980e51>:0                                       
  at System.Threading.Tasks.Task.ExecutionContextCallback (object) [0x00000] in <a9947af07e614fee8d8cc4b107980e51>:0                
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext,System.Threading.ContextCallback,object,bool) 
[0x00071] in <a9947af07e614fee8d8cc4b107980e51>:0                                                                                   
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext,System.Threading.ContextCallback,object,bool) [0x00000
] in <a9947af07e614fee8d8cc4b107980e51>:0                                                                                           
  at System.Threading.Tasks.Task.ExecuteWithThreadLocal (System.Threading.Tasks.Task&) [0x00050] in <a9947af07e614fee8d8cc4b107980e5
1>:0                                                                                                                                
  at System.Threading.Tasks.Task.ExecuteEntry (bool) [0x00058] in <a9947af07e614fee8d8cc4b107980e51>:0                              
  at System.Threading.Tasks.Task.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem () [0x00000] in <a9947af07e614fee8d8cc4b107980
e51>:0                                                                                                                              
  at System.Threading.ThreadPoolWorkQueue.Dispatch () [0x00074] in <a9947af07e614fee8d8cc4b107980e51>:0                             
  at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback () [0x00000] in <a9947af07e614fee8d8cc4b107980e51>:0              
  at (wrapper runtime-invoke) <Module>.runtime_invoke_bool (object,intptr,intptr,intptr) [0x0001e] in <a9947af07e614fee8d8cc4b107980
e51>:0                                                                                                                              
                                                                                                                                    
=================================================================                                                                   
Got a SIGSEGV while executing native code. This usually indicates                                                                   
a fatal error in the mono runtime or one of the native libraries                                                                    
used by your application.                                                                                                           
=================================================================                                                                   
                                                                                                                                    
Abort trap
```

**Comment by kenkendk** [comment]
Do you know what version of OpenSSL you have installed? I would prefer to know why it crashes, instead of defaulting to disabling it. If we can conclude that a specific version of OpenSSL causes the problem, I can either fix support for that version, or selectively disable it.

If you want to disable it on your system, you can set the environment variable `FH_LIBRARY=Managed`, which will switch over to using the previous version.

**Comment by buengese** [comment]
I'm allready using the environment variable.
I tried 2 different OpenSSL versions - both 1.0.1s-freebsd (the default version the system uses) and 1.0.2m-freebsd the most up-to-date version currently available as a package seem to crash.

**Comment by kenkendk** [comment]
Ok, I need to create a FreeBSD Docker image with OpenSSL installed before I can test it.

**Comment by kenkendk** [comment]
I just tried this with a clean `FreeBSD 11.1-RELEASE` image and did not get any crashes.

I only ran `pkg install mono` (and `pkg-install git` for getting the source code).

Summary:
```
pkg install mono
pkg install git
git clone https://github.com/kenkendk/FasterHashing
cd FasterHashing
xbuild FasterHashing.sln
cd FasterHashingTester/bin/Debug
mono FasterHashingTester.exe
```

**Comment by buengese** [comment]
Resolved with Mono 5.2.0.0.
